usr/lib/pkcs11/ep11_stdll/ep11_specific.c

```diff
@@ -5668,7 +5669,10 @@ CK_RV ep11tok_get_mechanism_list(STDLL_TokData_t * tokdata,
         *pulCount = 0;
         for (i = 0; i < counter; i++) {
             if (ep11tok_is_mechanism_supported(tokdata, mlist[i]) == CKR_OK) {
-                pMechanismList[*pulCount] = mlist[i];
+                if (*pulCount >= size)
+                    rc = CKR_BUFFER_TOO_SMALL;
+                else
+                    pMechanismList[*pulCount] = mlist[i];
```

Comment:
Id prefer to check *pulCount >= size directly after incrementing *pulCount below. If this is the case , set rc and break out of the for loop.
While you are at it, can you remove the empty else-case below ?

Reply:
Ah okay, you need to check it before, since the function does not check for pulCount=0 above.
So in this case i suggest to add a break statement to the if-case above, so you dont need an else-case anymore.

Reply:
*pulCount is the index to the array of mechs. After increasing it, you don't know if there will be a next mech (because all further might not be supported). If there is no next, then no need to set rc. 

Also if you return CKR_BUFFER_TOO_SMALL you must also return back the required number of elements in *pulCount. So we need to continue the loop to find out about the number of elements we really need.

I will remove the empty else.

Reply:
> So we need to continue the loop to find out about the number of elements we really need.

I was missing this. On first thought you dont see a reason to continue the for loop if its already clear, the function failed .. Can you do 
```
if (*pulCount < size)
    pMechanismList[*pulCount] = mlist[i];
```
and set the rc after the for loop. So this becomes more clear.

Reply:
Done.